### ISSUE TITLE
【1-1_load_vgg.ipynb】にてVGG-16のモデル構成が書籍と異なっていますか？

### ISSUE THREAD

**Issue by YutaroOgawa** [YutaroOgawa/pytorch_advanced#1]
対象ファイル：1-1_load_vgg.ipynb

VGG-16モデルをロードすると、書籍では「ネットワークは (features)と(classifier)の2つから構成されています。」と説明されていますが、 (features)と(classifier)の間に(avgpool)があり、3つから構成されています。

・第1章を進めるうえで、この違いは問題になりますでしょうか？
・この  (avgpool): AdaptiveAvgPool2d(output_size=(7, 7))は何でしょうか？

どうぞよろしくお願いいたします。

===（print(net)の出力結果）================
VGG(
  (features): Sequential(
    (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (1): ReLU(inplace)
  ・・・
    (29): ReLU(inplace)
    (30): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (avgpool): AdaptiveAvgPool2d(output_size=(7, 7))
  (classifier): Sequential(
    (0): Linear(in_features=25088, out_features=4096, bias=True)
    (1): ReLU(inplace)
    (2): Dropout(p=0.5)
    (3): Linear(in_features=4096, out_features=4096, bias=True)
    (4): ReLU(inplace)
    (5): Dropout(p=0.5)
    (6): Linear(in_features=4096, out_features=1000, bias=True)
  )
)
===================

**Comment by YutaroOgawa** [comment]
ありがとうございます。

書籍の説明との違いは、執筆時のPyTorchとtorchvisionのバージョンが現在の最新よりも古いために発生しております。
最新バージョンのtorchvisionで読み込むVGGモデルは、記載いただいた通り、モジュール数が3つになっています。

>第1章を進めるうえで、この違いは問題になりますでしょうか？

いいえ。問題ございません。この違いに起因するエラーはなく、第1章を進めることができます。

>この (avgpool): AdaptiveAvgPool2d(output_size=(7, 7))は何でしょうか？

このAdaptiveAvgPool2dは2次元のアベレージプーリング層です。output_sizeが7×7になっているのはfeatureモジュールからの出力サイズが7×7だからです。
「では、featureモジュールから7×7の特徴量が出てきたあと、7×7にアベレージプーリングするのであれば何の処理もしていないのと同じでは？」と思いますが、その通りです。
何の処理もしていません。

ただしこれは、入力画像をきちんと224×224にリサイズしてから、featureモジュールに入力しているためです。
featureモジュールに入力する画像のサイズが224×224でない場合、featureモジュールの出力は7×7にはなりません。
ですが、classifierモジュールは7×7の特徴量サイズを前提としているので、このような場合はエラーになってしまいます。

そこで、AdaptiveAvgPool2d(output_size=(7, 7))で、featureモジュールから出力される特徴量サイズが何であれ、7×7の特徴量サイズへと変換してからclassifierモジュールへと渡すことで、エラー発生を避けています。

このように、入力画像サイズに起因したエラーが起きないようにavgpoolモジュールが追加されている（と私は思っている）のですが、ともあれ、きちんと入力画像のサイズをリサイズしておくのが良いです。


----
参考：PyTorchのAdaptiveAvgPool2dの解説
https://pytorch.org/docs/stable/nn.html?highlight=adaptiveavgpool2d#torch.nn.AdaptiveAvgPool2d
